mesh-design.md

```diff
@@ -0,0 +1,270 @@
+and `setCollider` has yet another. This setup was done specifically to make it difficult to change actor
+colliders after actor creation. However, based on my research and with the exception of mesh colliders,
+creating a collider at patch time is no worse performance-wise than creating one at creation time. As such,
+the collider APIs should be consolidated to follow the same semantics as any other actor property.
+
+So instead of the `addCollider` flag:
+
+```ts
+MRE.Actor.CreatePrimitive(context, {
+	definition: {
+		shape: 'box',
+		dimensions: { x: 0.25, y: 0.25, z: 0.1 }
+	},
+	addCollider: true,
+	actor: { name: 'button' }
+});
+```
+
+The user would describe the desired collider in the patch:
+
+```ts
+MRE.Actor.CreateEmpty(context, {
+	actor: {
+		name: 'button',
+		appearance: {
+			meshId: assets.createBoxMesh('button', 0.25, 0.25, 0.1).id
+		},
+		collider: { geometry: { shape: 'auto' } as MRE.AutoColliderGeometry }
+	}
+});
+```
+
+This could also allow collider shape modification on the fly:
+
+```ts
+actor.collider.geometry = {
+	shape: 'sphere',
+	radius: 0.5
+} as MRE.SphereColliderGeometry;
+```
+
+I would also recommend removing `mesh` as a shape type, so it can only be specified implicitly
+via `auto`. Or else rename `mesh` to `unmanaged` so it's clear that it can't be used. Alternatively,
+we could go in the other direction and allow mesh colliders to be created at any time, and let
+the developer determine if it's appropriate to use.
+
+### Removing CreatePrimitive and CreateFromGltf
+
+Now that meshes can be created and assigned to actors after actor creation, these two functions
+are providing the same functionality (with reduced performance efficiency) as `Actor.CreateEmpty`
+with `AssetContainer.createPrimitiveMesh`, and `Actor.CreateFromPrefab` with `AssetContainer.loadGltf`.
+I propose that these functions be removed altogether, and some light syntax sugaring added to make
+their replacements more friendly.
+
+```ts
+// So this:
+MRE.Actor.CreatePrimitive(context, {
+	definition: {
+		shape: 'sphere',
+		radius: 0.5
+	},
+	addCollider: true,
+	actor: {
+		name: 'ball'
+	}
+});
+
+// Would become this:
+MRE.Actor.CreateEmpty(context, {
+	actor: {
+		name: 'ball',
+		appearance: {
+			meshId: assets.createSphereMesh('ball', 0.5).id
+		},
+		collider: { geometry: { shape: 'auto' } as MRE.AutoColliderGeometry }
+	}
+});
+
+// And this:
+MRE.Actor.CreateFromGltf(context, {
+	resourceUrl: url,
+	colliderType: 'mesh',
+	actor: {
+		name: 'gltf'
+	}
+});
+
+// Would become this:
+MRE.Actor.CreateFromPrefab(context, {
+	// new sugar: prefabId could take a promise that will resolve to an array of assets. Create and
+	// return the new local actor immediately, but postpone the create-from-prefab network message
+	// until the prefab promise resolves.
+	prefabId: assets.loadGltf(url, 'mesh'),
+	actor: {
+		name: 'gltf'
+	}
+});
```

Comment:
This is less discoverable than before. What about:
```
MRE.Actor.CreateFromGltf(this.assets, url, actor: { name: 'gltf' });
```

Reply:
It turns out this might be necessary regardless of my opinion on it. Turns out I can't arbitrarily postpone the message as I described here without also postponing every subsequent message that pertains to the created actor (set-behavior, set-sound-state, etc.). It's doable in the sync layer, but this seems controversial. Let's have a meeting to discuss.

---
In reply to: [311278822](https://github.com/microsoft/mixed-reality-extension-sdk/pull/409#discussion_r311278822) [](ancestors = 311278822)